src/contract/event.rs

```diff
@@ -14,64 +14,99 @@ use std::task::{Context, Poll};
-        }
-    }
+/// Additional metadata from the log for the event.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct EventMetadata {
+    /// The hash of the block where the log was produced.
+    pub block_hash: H256,
+    /// The number of the block where the log was produced.
+    pub block_number: u64,
+    /// The hash of the transaction this log belongs to.
+    pub transaction_hash: H256,
+    /// The block index of the transaction this log belongs to.
+    pub transaction_index: usize,
+    /// The index of the log in the block.
+    pub log_index: usize,
+    /// The log index in the transaction this log belongs to.
+    pub transaction_log_index: usize,
+}
 
+impl<T> Event<T> {
     /// Get a reference the underlying event data regardless of whether the
     /// event was added or removed.
-    pub fn as_data(&self) -> &T {
-        match self {
-            Event::Added(value) => value,
-            Event::Removed(value) => value,
+    pub fn inner_data(&self) -> &T {
+        match &self.data {
+            EventData::Added(value) => value,
+            EventData::Removed(value) => value,
         }
     }
 
-    /// Get a mutable reference the underlying event data regardless of whether
-    /// the event was added or removed.
-    pub fn as_data_mut(&mut self) -> &mut T {
-        match self {
-            Event::Added(value) => value,
-            Event::Removed(value) => value,
-        }
+    /// Gets a bool representing if the event was added.
+    pub fn is_added(&self) -> bool {
+        matches!(&self.data, EventData::Added(_))
+    }
+
+    /// Gets a bool representing if the event was removed.
+    pub fn is_removed(&self) -> bool {
+        matches!(&self.data, EventData::Removed(_))
     }
 
     /// Get the underlying event data if the event was added, `None` otherwise.
     pub fn added(self) -> Option<T> {
-        match self {
-            Event::Added(value) => Some(value),
-            Event::Removed(_) => None,
+        match self.data {
+            EventData::Added(value) => Some(value),
+            EventData::Removed(_) => None,
         }
     }
 
     /// Get the underlying event data if the event was removed, `None`
     /// otherwise.
     pub fn removed(self) -> Option<T> {
-        match self {
-            Event::Removed(value) => Some(value),
-            Event::Added(_) => None,
+        match self.data {
+            EventData::Removed(value) => Some(value),
+            EventData::Added(_) => None,
         }
     }
 }
 
+impl EventMetadata {
+    fn from_log(log: &Log) -> Option<Self> {
+        Some(EventMetadata {
+            block_hash: log.block_hash?,
+            block_number: log.block_number?.as_u64(),
+            transaction_hash: log.transaction_hash?,
+            transaction_index: log.transaction_index?.as_usize(),
+            log_index: log.log_index?.as_usize(),
+            transaction_log_index: log.transaction_log_index?.as_usize(),
```

Comment:
Looks like parity supports a log_type, which can be "mined" or "pending"https://github.com/andresilva/parity/blob/master/rpc/src/v1/types/log.rs. Doesn't seem part of the RPC standard but could be interesting to optionally support nonetheless?